src/java/htsjdk/samtools/SamPairUtil.java

```diff
@@ -426,27 +434,35 @@ else if (record.getSecondOfPairFlag()) {
                 }
                 records.add(record);
             }
+            // TODO: should we check that we do not have a mix of paired and fragment reads?
+
 
             // we must find both records to update the mate info
             if (null != firstPrimaryRecord && null != secondPrimaryRecord) {
                 // Update mate info
                 SamPairUtil.setMateInfo(firstPrimaryRecord, secondPrimaryRecord, this.setMateCigar);
                 if (this.setMateCigar) this.numMateCigarsAdded += 2;
-            }
-
-            // Set mate information on supplemental records
-            if (containsSupplementalRecord) {
-                for (final SAMRecord record : records) {
-                    if (record.getReadPairedFlag() && record.getSupplementaryAlignmentFlag()) {
-                        if (record.getFirstOfPairFlag()) {
-                            SamPairUtil.setMateInformationOnSupplementalAlignment(record, secondPrimaryRecord, this.setMateCigar);
-                        }
-                        else {
-                            SamPairUtil.setMateInformationOnSupplementalAlignment(record, firstPrimaryRecord, this.setMateCigar);
+                //}
+
+                // Set mate information on supplemental records
+                if (containsSupplementalRecord) {
+                    for (final SAMRecord record : records) {
+                        if (record.getReadPairedFlag() && record.getSupplementaryAlignmentFlag()) {
+                            if (record.getFirstOfPairFlag()) {
+                                SamPairUtil.setMateInformationOnSupplementalAlignment(record, secondPrimaryRecord, this.setMateCigar);
+                            } else {
+                                SamPairUtil.setMateInformationOnSupplementalAlignment(record, firstPrimaryRecord, this.setMateCigar);
+                            }
+                            this.numMateCigarsAdded++;
                         }
-                        this.numMateCigarsAdded++;
                     }
                 }
+            } else if (!this.ignoreMissingMates) {

```

Comment:
I think this could still fail with an NPE in the (admittedly invalid edge) case where *all* you have is supplemental reads and both the R1 and R2 primary records are missing.